src/test/java/uk/gov/hmcts/reform/idam/web/AppControllerTest.java

```diff
@@ -1933,4 +1933,79 @@ public void verificationView_shouldPopulateAuthorizeCommand() throws Exception {
             .andExpect(model().attribute(SELF_REGISTRATION_ENABLED, Boolean.TRUE))
             .andExpect(view().name(VERIFICATION_VIEW));
     }
+
+    /**
+     * @verifies validate code field is not empty
+     * @see AppController#verification(uk.gov.hmcts.reform.idam.web.model.VerificationRequest, BindingResult, Model, HttpServletRequest, HttpServletResponse)
+     */
+    @Test
+    public void verification_shouldValidateCodeFieldIsNotEmpty() throws Exception {
+        mockMvc.perform(post(VERIFICATION_ENDPOINT).with(csrf())
+            .cookie(new Cookie("Idam.AuthId", "authId"))
+            .header(X_FORWARDED_FOR, USER_IP_ADDRESS)
+            .param(USERNAME_PARAMETER, USER_EMAIL)
+            .param(REDIRECT_URI, REDIRECT_URI)
+            .param(STATE_PARAMETER, STATE)
+            .param(RESPONSE_TYPE_PARAMETER, RESPONSE_TYPE)
+            .param(CLIENT_ID_PARAMETER, CLIENT_ID)
+            .param(SCOPE_PARAMETER, CUSTOM_SCOPE)
+            .param(CODE_PARAMETER, ""))
+            .andExpect(status().is2xxSuccessful())
+            .andExpect(view().name(VERIFICATION_VIEW))
+            .andExpect(model().attribute("isCodeEmpty", true));
+
+        verify(spiService, never()).submitOtpeAuthentication(any(),
+            any(),
+            any());
+    }
+
+    /**
+     * @verifies validate code field is digits
+     * @see AppController#verification(uk.gov.hmcts.reform.idam.web.model.VerificationRequest, BindingResult, Model, HttpServletRequest, HttpServletResponse)
+     */
+    @Test
+    public void verification_shouldValidateCodeFieldIsDigits() throws Exception {
+        mockMvc.perform(post(VERIFICATION_ENDPOINT).with(csrf())
+            .cookie(new Cookie("Idam.AuthId", "authId"))
+            .header(X_FORWARDED_FOR, USER_IP_ADDRESS)
+            .param(USERNAME_PARAMETER, USER_EMAIL)
+            .param(REDIRECT_URI, REDIRECT_URI)
+            .param(STATE_PARAMETER, STATE)
+            .param(RESPONSE_TYPE_PARAMETER, RESPONSE_TYPE)
+            .param(CLIENT_ID_PARAMETER, CLIENT_ID)
+            .param(SCOPE_PARAMETER, CUSTOM_SCOPE)
+            .param(CODE_PARAMETER, "a"))
+            .andExpect(status().is2xxSuccessful())
+            .andExpect(view().name(VERIFICATION_VIEW))
+            .andExpect(model().attribute("isCodePatternInvalid", true));
+
+        verify(spiService, never()).submitOtpeAuthentication(any(),
```

Comment:
Using `any()` on verify might become a problem later with mutation tests.

Reply:
this test is checking this feature is never called, so shouldn't make a difference